---
title: "Lab3 - Single Cell RNAseq"
author: "Original : Joseph Bergenstr√•hle - Edit : Alma Andersson, Sailendra Pradhananga "
date: "06 Dec 19"
output:
  tufte::tufte_html: default 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=FALSE)
```

____
## Lab3 - Single cell RNA seq (scRNA-seq)
____

Today, it's possible to get genomic information even from single cells. (You
should have heard about single cell RNA-seq in the lectures!). Traditionally,
many of the methods used to analyze RNA-seq data were developed with standard
"bulk" RNA-seq in mind. However, some of the underlying assumptions that are
used for "bulk" RNA-seq methods are not appropriate for analyzing scRNA-seq
data.

____
## 1) Introduction
____

Short primer: what is scRNA-seq?

*  Measures the expression levels for each gene across a population of cells.
* Adresses biological questions in which cell-specific changes in
  transcriptome are important, e.g. cell type identification, heterogeneity of
  cell responses, stochasticity of gene expression, inference of gene regulatory
  networks across the cells.
* Currently there are several different protocols in use, e.g. SMART-seq2,
  CELL-seq and Drop-seq.
*  There are also commercial platforms available, including the Fluidigm C1, 10X
   Genomics Chromium etc.
*  Several computational analysis methods from bulk RNA-seq can be used.
*  In most cases computational analysis requires adaptation of the existing
   methods for bulk RNA-seq analyses or development of new ones.

This is a general pipeline for an scRNA-seq analysis workflow

![](flowchart_rnaSeq.png)

____
## 2) Construct the expression matrix
____

Ok, let's start by looking at our expression matrix. Note that in the workflow
schematic above, we will skip the Read QC, Alignment, Mapping QC, and Cell QC
steps. Let's assume this (boring ;-)) stuff has already been done.

Here we have provided you with an already quality-controlled dataset, consisting
of pluripotent stem cells. Load the data!

```{r, eval=FALSE}
data <- read.table("data/tung.tsv")
```

The object called `data` is an object of class `data.frame`. A `data.frame` 
object is usually used to store any kind of tabular data in R. You can check 
the dimensions of the `data.frame` to count the number of columns and rows.

```{r, eval=FALSE}
# Check class
class(data)

# Check dimensions
dim(data)
```

### The Seurat R package 

We will use the "Seurat" R package, developed by the Satija Lab
https://satijalab.org/ over at NYU. Seurat has become one of the most popular 
tools for single cell analysis in R and is constantly updated with new features. 
There are many tutorials available on their website to get started with various 
types of analyses which is also a reason why it has become so popular among 
researchers.

A core feature of Seurat is the "Seurat" object. Without going into too much detail,
this is an object designed to store all relevant input data as well as analysis results.

If you have installed the Seurat R package, the only thing you need to create a 
"Seurat" object is an expression matrix with cells in columns and genes in rows. 
Once this object has been created, Seurat provides all the tools necessary to run 
the steps described in the flowchart above using a set of predefined functions. 

Load the Seurat R package (install via BiocManager::install() if not already installed)

```{r, eval=FALSE}
#BiocManager::install('Seurat')
library(Seurat)

```

Before we create our Seurat object, it can be useful to include some meta data as well. 
This is a table with additional information about our cells, e.g. if they were processed
in separate batches or obtained from different individuals. 

```{r, eval=FALSE}
metadata <- read.table(file = "data/tung_metadata.tsv", stringsAsFactors = FALSE)
head(metadata)
```

Note that the rownames of the meta data table matches the column names of the expression
matrix. This is important to make sure that the meta data is associated with the correct cells.

Let's create a Seurat object with our expression matrix and meta data table.

```{r, eval=FALSE}
se <- CreateSeuratObject(counts = data, meta.data = metadata)
```

To access the meta data from your new Seurat object, you can use double brackets `[[]]`.

```{r, eval=FALSE}
head(se[[]])
```

Notice that you have some additional columns in your meta data whcih have been added 
automatically when you created your Seurat object. For example; nCount_RNA which is the number 
of unique molecules detected per cell and nFeature_RNA which is the number of unique genes
detected per cell.

You can access any column from you Seurat object using the `$`symbol, just like you would
with a `data.frame` object. Below is an example of how to extract the "replicate" column
and then summarizing how many cells from each replicate that we have.

```{r, eval=FALSE}
table(se$replicate)
```

*Q1:* How many genes do we have? How many cells?

```{r, eval=FALSE, echo=TRUE}
# Write your code / answer here
```

Single-cell data typically contains _a lot_ of "samples" (cells). Due to the
size of the count matrix and the high frequency of 0 values, Seurat stores the
data in a "sparse matrix format" to save memory. 

You can access the raw data using the `GetAssayData` function and specify that 
you want to collect the raw expression matrix by typing `slot = "counts"`. Note 
that this matrix is the same as the `data` matrix that we used to create our Seurat object, 
but it's important to know how to extract data from the Seurat object as well. For example if 
you want to access this matrix to apply a method which is not available from the 
Seurat R package. 

```{r, eval=FALSE}
sparse.exprMat <- GetAssayData(se, slot = "counts")
```

Now we have the raw expression matrix in a sparse format saved to a variable called 
`sparse.exprMat`. Let's can have a look at the top left corner of the matrix (first 5 genes
and first 5 cells). 

```{r}
# Check class of object 
class(sparse.exprMat)

# Extract top left corner
sparse.exprMat[1:5, 1:5]
```

Note that there are a lot of dots in the expression matrix which represent 0 values and
the class is called "dgCMatrix" which is a type of sparse matrix format. 

Column names such as "NA19098.r1.A01" or "NA19098.r1.A02" represent unique cell identifiers,
whereas the gene names are given here as gene symbols (HGNC nomenclature).

Sometimes, it's easier to work with a "normal" matrix and if you want to you can simply convert
the sparse matrix into a normal matrix using the `as.matrix` function:

```{r, eval=FALSE}
exprMat <- as.matrix(sparse.exprMat)

# Check class of object 
class(exprMat)

# Extract top left corner
exprMat[1:5, 1:5]
```

Note that the 0 values are back when vieweing the "normal" matrix. Many functions 
used in R are specifically written to manipulate certain classes of objects and for 
this reason you have to be careful and make sure that you have have the correct input 
for the function that you want to use. Keep this in mind whenever you get an error message!

____
## 3) Quality Control (QC)
____

Now, some cells in your expression matrix mght be too bad to include in your analysis. 
Cells with very few detected RNA molecules might be too sparse to extract meaningful 
information from. However, if these cells are kept they could have a negative impact on our
analysis. It's usually a good idea to explore the data visually to make sense of various
quality aspects and then apply some filters to decide what cells to keep.

### Total number of reads obtained

To visualize the distribution of total number of reads per cell, we can use a histogram. 
Take a look at this pre-loaded example data showing air temperatures registered over 9 months:

```{r, eval=FALSE, echo=TRUE}
# Extract the column called Temp and save it to a new variable
Temperature <- airquality$Temp

# Plot histogram
hist(Temperature)
```

We can add additional things to our histogram, e.g. a vertical line
highlighting a cut-off point:

```{r, eval=FALSE, echo=TRUE}
# plot histogram
hist(Temperature)

# Add a vertical line
abline(v = 73, col = "red")
```

*Q2:* Make a histogram of the total number of RNA molecules detected per sample.
Based on prior experience of working with scRNA-seq data and the shape of the
distribution, you decided of filtering threshold of 1 million reads. Include a
vertical line in your figure to indicate this threshold.

Hint: you can access meta data using the `$` symbol

```{r, eval=FALSE, echo=TRUE}
# Write your code / answer here
```

### Nr of unique genes detected

In addition, to ensure a sufficient sequencing depth for each sample, we also
want to make sure that the reads are evenly distributed across the
transcriptome.

*Q3:* Make the similar kind of histogram as before, but now show the distribution
of the number of genes detected within each sample. This time, you decide to use
a filtering threshold of 7000 genes. Include this threshold as a vertical line
in your figure.

```{r, eval=FALSE, echo=TRUE}
# Write your code / answer here
```

_____
## 4) Filter
_____

Filtering a "Seurat" object from low quality cells can be done by first identifying 
the cell names that you want to keep and use these names to subset the object.

Let's filter the cells that fall below the thresholds from the earlier
questions, i.e. with less than 1 million unique molecules or less than 7 thousand 
unique genes. Below are a few lines of code to extract cell names for cells meeting 
these two criteria and these cell names are stored in the `keep.cells` character vector.

*Q4:* Use the `keep.cells` character vector to subset the "Seurat" object and complete the
code chunk below. Save the new "Seurat" object into a variable called `se.subset`.

Hint: You can for example use the `subset`. Type ?subset.Seurat in the R console to 
open up a documentation window for the subset function.

```{r, eval=FALSE, echo=TRUE}

# Select cell names with at least 1 million unique molecules
set1 <- colnames(se)[se$nCount_RNA >= 1e6]
# Select cell names with at least 7000 unique genes
set2 <- colnames(se)[se$nFeature_RNA >= 7000]

# Select cells which meet both criteria above (intersect)
keep.cells <- intersect(set1, set2)

# Now use these cell names to subset the "Seurat" object

# Write your code / answer here

```

You will often find that some genes are lowly expressed across your dataset. It can be a 
good idea to filter out such genes as these are not always informative. We can compute the 
overall expression per gene by summing up the count for each row:

```{r, eval=FALSE}
# First get the expression matrix
sparse.exprMat <- GetAssayData(se, slot = "counts")

# Convert it into a regular matrix
exprMat <- as.matrix(exprMat)

# Calculate row sums
gene.counts <- rowSums(exprMat)

# Check the results 
head(gene.counts)
```

*Q5:* Compute a new vector (call it `keep.genes`) containing the genes with a total
expression lower than 500 across the whole dataset. The `keep.genes` character vector 
should consist of genes symbols.

Hint: You can use the vector called `gene.counts` to determine what genes to keep. 
Check the example above where we filtered out cells.

```{r, eval=FALSE, echo=TRUE}
# Write your code / answer here
```

Now that we have this vector we can subset the "Seurat" object we did previously,
but this time we'll apply the filter to the genes instead.

```{r, eval=FALSE, echo=TRUE}
se.subset <- subset(se.subset, features = keep.genes)
```

Now that we have filtered our object (`se.subset`), we should have fewer cells
and fewer genes. 

*Q6:* How many cells do we have now? How many genes?

```{r, eval=FALSE, echo=TRUE}
# Write your code / answer here
```

_____
## 5) Normalization
_____

There are some specific challenges associated with single cell data. In particular, 
the major challange is that the data is much sparser compared to bulk RNA-seq data. 
A result of this is that the number of genes detected in a single cell is typically 
only a fraction of the number of genes expressed across the genome. In other words, 
a lot of genes have 0 expression in a cell.

The missing genes are referred to as **gene "dropouts"** and makes it challenging to 
compare expression levels across cells. 

Dropouts are less common in bulk sequencing data because you measure expression 
levels from a large number of cells at the same time. 

The total amount of RNA molecules that can be detected in a cell (referred to as "library size") 
depends on a number of technical and biological features. For example, larger cells typically 
contain more RNA molceules while cells with a lot of RNAses contain less RNA as a result of degradation. 
These are examples of sources of variability which we are not interested in and are commonly 
referred to as "confounding factors". 

To make the expression levels comparable across cells, we can adress confounding factors
by applying a normalization method. A good normalization strategy can be used to transform a
raw expression matrix into a new matrix where expression levels are comparable across cells 
in a meaningful way. You can find a number of normalization 
strategies and more in depth theory [here](https://scrnaseq-course.cog.sanger.ac.uk/website/cleaning-the-expression-matrix.html#normalization-theory).

In this course, we will use a method called variance stabilizing transformation, 
[VST](https://www.biorxiv.org/content/10.1101/576827v2) method which is provided 
by the `SCTransform()` function from the Seurat package:

```{r, eval=FALSE}
se <- SCTransform(se)
```

Now, you will have a new "assay" stored in the Seurat object called "SCT",
which contains the normalized data. This assay will automatically become 
the "default assay" for subsequent analyses. The "active" assay contains 
data that it will be used as input to any method that you apply to your 
Seurat object. 

If you type `DefaultAssay(se)`, you should get "SCT".

```{r}
DefaultAssay(se)
```

If you just print the object in your console, you will get some basic information 
about the dataset such as what assays that are present and which one is active. 
You could have any number of assays available and it's important to know what assay
you're working on.

```{r}
se
```

_____
## 6) Dimensionality reduction
_____

A single-cell dataset is composed of several cell types which are all characterized by the 
expression of specific sets of genes. If we were to zoom in on a specific cell type, it is 
reasonable to expect that there exists some set a co-expressed genes which defines this cell type. 
For example, you might find that a cell type express genes related to neuronal activity, immune 
respones or oxygen transport. 

In reality, an expression matrix represent many complex interactions between different genes and these 
interactions can vary across different cell types and cell states. But there are typically some very clear
patterns in the data and what we want to do is to make sense of these patterns. 
One major benefit with scRNA-seq data is that we can fish out different patterns using machine learning 
methods. Dimensionality reduction methods represent a family of machine learning methods which are 
particularly suitable for this task.

As the name implies, a dimensionality reduction method can be used to reduce a high dimensional dataset into a 
representation with fewer dimensions. These new dimensions can pick up sources of variability in the dataset, 
such as signalling pathways, immune functions, metabolic activity, apoptotic activity and much more. The most 
useful patterns are often the ones that represent some biological variability, but you can also find patterns 
that represent technical variability. An example of this could be if you have doublet cells in your data. We are 
only interested in analyzing individual cells, but sometimes two or more "connected" cells slip through the 
experimental procedure because of technical limitations. We can adress most of the technical variability using 
normalization and filtering procedures, but you should always keep an eye things like this.

The interpretation of dimensionality reduction vectors can be confusing, so it can be helpful to think of them as 
patterns representing some biological or technical variability.

For scRNA-seq data, dimensionality reduction methods are also crucial to visualize the data! 

First, let's have a look at the most widely used dimensionality reduction method called 
Pincipal Component Analysis (PCA). This method can be applied to our Seurat object 
using the `RunPCA` function.

Let's first compute a PCA representation of our data:

```{r, eval=FALSE}
se <- RunPCA(se)
```

We can then plot the transformed data by using the function `DimPlot()` and specify the reduction that we want to use:

```{r, eval=FALSE}
DimPlot(se, reduction="pca", dims = c(1, 2))
```

Here we have plotted the first two PCA vectors (1 and 2) by setting `dims = c(1, 2)` 
and you can only visualize two at the time. PCA vectors are scaled at 0 so you will always 
get both positive and negative values on the axes. You can already see that the cells 
form some sort of structure in the plot.

We can also color the cells using information from our meta data table. Let's color
the cells based on indiviudal:

```{r}
DimPlot(se, reduction = "pca", dims = c(1, 2), group.by = "individual")
```

Now we can see that there is some separation of the cells based on individual. Exploring these types
of plots can be useful to find out if variablity in your data can somehow attributed to previous
knowledge about your data. Looking at PC_2 you can see that a population of cells from individual 
NA19098 differ from the other two individuals. 

*Q7:* Produce a PCA plot where data points have been colored by "individual" and shaped by "replicate".

Hint: Read the help text on `DimPlot()` and, specifically, the "shape.by" and "group.by" arguments!

```{r, eval=FALSE}
# Write your code / answer here
DimPlot(se,
        reduction = 'pca',
        group.by = 'individual',
        shape.by = 'replicate')

```

*Q8:* Try some other dimensionality reduction methods and play around with
coloring the cells, can you draw any conclusion from the data?

Note that Seurat comes with a bunch of other methods for dimensionality reduction such as 
`RunICA()`,`RunUMAP()`, `RunTSNE`, ...

```{r, eval=FALSE}
# Write your code / answer here
```

For the rest of this lab, we'll use a different dataset with peripheral
blood mononuclear cells (PBMCs). PBMCs are extracted from blood samples 
and contain mainly immune cells such as lymphocytes and monocytes. 

When you run `pbmc <- readRDS("data/pbmc")` 
you will get a Seurat object that has already been normalized. 

```{r}
pbmc <- readRDS("data/pbmc")
```

Let's try another dimensionality reduction method called UMAP. This method
is commonly used to "embed" the cells into a 2D map that can be very useful
for visualization purposes. 

```{r, eval=FALSE}
pbmc <- RunUMAP(pbmc, dims = 1:30)
DimPlot(pbmc, reduction = "umap")
```

Here you can already see some interesting structures in the data; some bigger blobs and some
smaller blobs. This type of visualization can be very useful to get some overview of the data
but we shouldn't draw too many conclusions from it. Now the question is, where do we draw the 
lines to separate these blobs into groups of similar cells?

_____
## 7) Clustering 
_____

Clustering of scRNA-seq data refers to methods used to group samples (cells) based on gene expression
similarity (https://www.nature.com/articles/nmeth.4299). The methods are typically 
unsupervised or semi-supervised, meaning that we can define group structures in the 
data with little or no prior knowledge about the data beforehand. It's important to 
keep in mind that clustering methods will find structure in the data, but doen't let 
you know if the structures make sense or not. 

In the case of scRNA-seq analysis, you are typically interested in defining cell types. 
What people usually do is to cluster the cells and then define (annotate) the clusters as 
different cell types based on their specific expression of "marker genes". Here, marker 
genes can be thought of as genes known to be specifically expressed by a specific cell type, 
e.g. CD19 is known to be expressed only by B-cells. However, you should know that most of the 
time you can tweak parameters to obtain different clusters and it can be very difficult to 
find the optimal parameter settings. 

Let's try clustering our data using the default method provided by Seurat's 
`FindNeighbors()` and `FindClusters()` functions. The parameter called `resolution` can
be tweaked to obtain a different number of clusters. Lowering this value will result in fewer clusters
and increasing this value will result in more clusters. Here we will use the default resolution of 0.8.

*Q9:* Use `DimPlot` to color the cells by cluster on the "umap" graph. Write down how many 
clusters you got using a resolution of 0.8.

Hint: Look at the meta.data of you Seurat object. Here you can find the clustering results
which will be called something like "SCT_snn_res...".

```{r, eval=FALSE}

pbmc <- FindNeighbors(pbmc, dims = 1:30, reduction = "pca")
pbmc <- FindClusters(pbmc, resolution = 0.8)

# Write code here

```


*Q10:* Re-run the `FindClusters()` function using a resolution of 0.3 and 1.2. Use `DimPlot` 
to visualize these clusters on the "umap" graph and write down how many clusters you get.

```{r, eval=FALSE}
# Write your code / answer here
```

We have already provided a column in the meta data slot with celltype labels so you do 
not need to do the dirty work of defining clusters, identifying markers and annotating 
the clusters.

```{r}
pbmc <- SetIdent(pbmc, value = "celltype")
DimPlot(pbmc, group.by = "celltype", label = T)
```

These cluster annotation have been determined based on marker genes found in the literature.
For example, MS4A1 is expressed specifically by B-cells, PPBP on platelet cells and CD14 on 
Monocytes. 

*Q11:* Draw a new plot showing the expression of these three markers across the celltypes. 

Hint: You can use the function `FeaturePlot()` (type `?FeaturePlot? in the R console  for details) which is very 
similar to the function `DimPlot()`. Alternatively you can visualize the marker expression 
using another function called `VlnPlot()` (type ?VlnPlot in the R console for details).

```{r, eval=FALSE}

# Write your answer here
```

_______
## Hand In Guidelines
_______
**How :** Briefly answer the questions marked with a ''Q'' in .Rmd file. There
are 11 of those. You can your answer in this .Rmd file and submit your complete
''.Rmd'' as your ''report'' for Lab2.

**Deadline**: Your report is due 23:59 December 13th 2020; if working in pairs -
each of you should hand in (identical) reports where the names of both authors
are clearly stated. For further information see the guidelines on the course
web-page.
