#+REVEAL_ROOT: ./reveal.js
#+REVEAL_TRANS: slide
#+REVEAL_EXTRA_CSS: ./center.css
#+OPTIONS: num:nil toc:nil
#+REVEAL_THEME: night 
#+MACRO: color @@html:<font color="$1">$2</font>@@
#+HTML_HEAD: <base target="_blank">


#+Title: CB2040 Lab Prep
#+Author: Alma Andersson
#+Email: almaan@kth.se

* Lab course Introduction
  
For these slides: [[https://almaan.github.io/genetech/][https://almaan.github.io/genetech/]]
  
** Teaching assistants   

There are 3 of us, all working at the dept. of Gene Technology    

  - Alma Andersson (almaan@kth.se)
  - Ludvig Larsson (ludvig.larsson@scilifelab.se)
  - Sami Saarenpää (sami.saarenpaa@scilifelab.se)
 
  [[./images/tas.png]]

Emails are welcome, but hands-on help with the exercises will be limited to the lab
sessions.

**NOTE**: Tag emails with [CB2040] in the subject line.\\
For example: /"[CB2040] Question about lab 2"/

** Setup

There are 4 labs in total, one "main TA" for each:
#+ATTR_REVEAL: :frag (fade-in fade-in fade-in fade-in) :frag_idx (1 2 3 4 )
   1. Introduction to R [Alma]
   2. Bulk RNA-seq analysis [Sami]
   3. Single Cell RNA-seq analysis [Ludvig]
   4. Spatial Transcriptomics analysis [Alma]

All labs are in R, we assume that you have:
#+ATTR_REVEAL: :frag (fade-in) : frag_idx(5 6 )
- access to a computer
- **no prior** knowledge of R
- some familiarity with the {{{color(lightgreen,terminal)}}}
 
** Formalities 
 
*Workflow:*
#+ATTR_REVEAL: :frag (fade-in) :frag_idx(1 1 1)
  * Labs are written in R markdown (more on this later)
  * Download exercises --> modify files in-place --> hand in modified file
  * You hand in the modified file via **Canvas**

*Deadlines:*
#+ATTR_REVEAL: :frag (fade-in) :frag_idx(2 2 2)
    - Hand-in: 1 week after lab session
    - Grading : 1 week after hand-in deadline
    - Revision : 1 week after correction deadline
 
#+ATTR_REVEAL: :frag (fade-in) :frag_idx(3)
/NOTE:/ Late reports will be corrected, but you will get a set of
additional exercises to solve. An incomplete report
is also considered late.
 
*Working in pairs:*
#+ATTR_REVEAL: :frag (fade-in) :frag_idx(3 3 3 3)
    - OK!
    - Pair = 2 people
    - hand in /identical/ copies
    - both names should be stated in file
    - state in comment who you worked together with

** Finding Information
   All information you need can be found at:\\
#+ATTR_HTML: :target "blank"
[[https://github.com/almaan/genetech][https://github.com/almaan/genetech]]

- I will refer to this as "the course repo".
- Bookmark or save the link, we will be using it today

* R
  
** What?
#+ATTR_REVEAL: :frag (fade-in) 
   - Programming language for "statistical computing"
   - Released in 1993
   - Open Source --> we like free stuff!
   - More than 15,000 additional packages! (add-ons)
   - Interpreted language --> very easy to work with!

#+REVEAL_HTML: <ul><li class="fragment fade-in visible" style="list-style-type:none;">
   Example C++ : Compiled language
#+BEGIN_SRC c++
    std::vector<double> vec1 = {1,2,3,4};
    std::vector<double> vec2 = {5,6,7,8};
    double prod = 0;

    for (int i = 0; i<= vec1.size(); i++){
      prod += vec1[i]*vec2[i];
    }
#+END_SRC
Example R : Interpreted language
#+BEGIN_SRC R
  vec1 <- c(1,2,3,4)
  vec2 <- c(5,6,7,8)
  prod <- vec1 %*%vec2
#+END_SRC
#+REVEAL_HTML: </li></ul>
    
** Why?
#+ATTR_REVEAL: :frag (fade-in fade-in fade_in fade-in fade-in) :frag_idx (1 2 3 4 5)
   - One the most popular languages within the bioinformatic community
   - Extremely well designed for "/data wrangling/"
   - Has a **huge** ecosystem of packages for bioinformatic analysis
   - Superior graphics system (ggplot) - don't underestimate this...
   - We believe it will give you an attractive and useful set of skills
   
   @@html:<span style="font-size:14px;font-style:italic">Sidenote: I hate R...</span>@@
  
** Installation
We will start by installing R, to do this:
  1. Goto the course repo
  2. Find the section "R - a primer"
  3. Click the link listed after "[[https://ftp.acc.umu.se/mirror/CRAN/][Download R]]"
  4. Choose the proper OS and follow the instructions
 
To test if you have a successfull install:
Open up the terminal and type "R", then press ENTER\\
\\
You should see something like:
#+BEGIN_SRC
almz :: teaching/genetech » R

R version 3.6.1 (2019-07-05) -- "Action of the Toes"
Copyright (C) 2019 The R Foundation for Statistical Computing
Platform: x86_64-redhat-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> 
#+END_SRC
 
* Rstudio
  
** What is it?
 
#+ATTR_REVEAL: :frag (fade-in) 
- When editing code it's common to use an IDE (/Integrated Development Environment/)
- IDEs are designed to facilitate programming; there are plenty of them
- Personal favorite is Emacs (first release 1976)
- We will be using {{{color(orange,Rstudio)}}}, an IDE specifically designed for R; also very user friendly
- In short, Rstudio is the tool you will be using to /edit/ your code.

** Installation
   To install Rstudio, follow the steps below:
   
    1. Head back to the course repo and the "/R - a primer/" section
    2. Click the link listed after "[[https://rstudio.com/products/rstudio/download/#download][Download Rstudio]]"
    3. Choose the proper OS and follow the instructions
    4. Shout at me if you experience issues!
\\
If you have a successful install, the Rstudio program should be available for you to open. The icon looks something like:\\
#+ATTR_HTML: :width 40%
[[https://helloacm.com/wp-content/uploads/2017/10/R-studio.png]]
\\
If you don't have a GUI - from the terminal, type in ~rstudio~ and the program should open up

** Customizing Rstudio

To get an idea of how Rstudio works, we'll have a quick look at it.\\
\\
So, **Open up Rstudio**.\\
\\
There are a lot of things you can customize, for example: /color schemes, fonts, pane layouts, keybindings etc./
\\
To change any of these : /Tools-->Global Options/

*** Pane Layout
    You may use whatever settings you want, but I recommend - to make things
    easy - that you use the same pane layout as me:
#+ATTR_HTML: :width 40%
    [[./images/pane-layout.png]]

** Orienting Rstudio
   - Create a new file by: /File --> New File --> R Script/
   - You should see something like this:
   #+ATTR_HTML: :width 60%
   [[./images/rstudio-show.png]]
   - There are some key features I want to point out: **{{{color(cyan, Source\, Console\, Terminal\, and Environment)}}}**
   - One more feature (/knitting/) that I will come back to

* The Labs
  
** Accessing the material
 
**Git and GitHub**\\
    - All lab material is hosted at github (the course repo).
    - **Git** is a system for distributed version control - github is one (nice) interface
    - github is widely used among bioinformaticians and computational biologist. I host all my software there
\\
**Download**\\
To create a local copy of the labs on your own computer:
1. Open a terminal window
2. Change to a directory  where you want the copy to be created
3. Run the command:
#+BEGIN_SRC bash
git clone https://github.com/almaan/genetech.git
#+END_SRC

You should now have a folder called **genetech** in your directory. Make sure such is the case.

** Installing packages
 \\  
   - Packages allows you to execute complex actions without the need to write 1000s of lines of code --> NICE!
   - Dependencies between packages can cause **a lot** of errors --> HEADACHES..
   - I have written an installation script that should take care of this for you (fingers crossed..)

\\
  1. In a terminal, go to /"=genetech/labs/pre="/
  2. Execute the following command:
#+BEGIN_SRC R
Rscript --vanilla ./install-packages.R
#+END_SRC

** Working with the code

To get an idea of how you work with the code, we will have a look at Lab 1
   - Go to Rstudio
   - Do: /File-->Open File/
   - Choose the file /"=genetech/labs/ex1/main.Rmd="/

*** Changing author name
    
  At the top of the file you will see:

  #+BEGIN_SRC yaml
  ---
  title: "Lab 1 - Introduction to R"
  author: "Alma Andersson"
  date: "11-09-2020"
  output:
    tufte::tufte_html: default 
  ---
  #+END_SRC

Change the author field to your name(s)

*** Code chunks
    \\
#+BEGIN_SRC rmd    
{r,eval = TRUE, echo=TRUE}
# bind the values 1337 to the variable fancy_name_1
fancy_name_1 <- 1337
# print fancy_name_1
print(sprintf("fancy_name_1 is : %d",fancy_name_1))
#+END_SRC
 #+ATTR_REVEAl: :frag (fade-in)
    - The "boxes" with code are called {{{color(orange,chunks)}}}.
    - If you "run" a chunk (press green arrow) the code inside will be exectueted
    - This is where /you/ will put in code (when prompted)
    - Don't change the settings (in curly brackets)


*** Knitting
 #+ATTR_REVEAL: :frag (fade-in)
    - R markdown --> {easy} --> html and pdf documents.
    - This process is called /knitting/
    - To knit your document, press the **yarn symbol** that says "**Knit**"
    - Your rendered document should show up in the "/View/" pane
    - You should **always** hand in a knitted report (html)
    - If your document does not knit successfully, something is wrong.

*** Knitting - a demo
    @@html:<video controls src="./vids/knit.mp4"></video>@@
      
* Wrap up
  - First lab session: 13:00-17:00, 9/11/2020 @ B3 ~(59°21'05.7"N 18°04'06.8"E)~
  - Questions are welcome!
  - I will hang around a bit after this session.
   
@@html:<h2> Thanks for the attention!</h2> @@
#+ATTR_HTML: :width 25%
[[https://img2.pngio.com/download-hd-owl-burrowing-owl-transparent-png-image-nicepngcom-burrowing-owl-png-386_345.png]]
